### PR TITLE
Use graph colouring to assign routing keys

### PR DESCRIPTION
**Pull request by mundya**
Uses graph colouring to assign identifiers to signals (and hence keys to multicast packets). This reduces the number of identifiers that are required (previously each net had a unique identifier) and consequently the number of bits in the key that are required to represent the identifier.

The results of this are that:
- The number of bits required is less strongly tied to the number of Signals in the model  -- allowing larger models to be simulated without running out of bits in the key (necessary for Spaun)
- Routing tables might be slightly smaller

Since this is a first pass there is scope for improvement. In particular it is currently required that all of the multicast routes which arrive at the same Nengo object have different keys - in principle this constraint can be relaxed such that the same keys may be used by any nets arriving at the same object given that (a) the nets do not route differently from the same SpiNNaker node and (b) they do not target different ports or filters in the object. Adjusting this should allow the size of the filter routing table to be reduced, resulting in a further performance improvement.

_Headline figure_ in a model of the 512-d circular convolution net (as instantiated on SpiNNaker) this technique reduced the number of identifiers from ~10 000 to ~900 or 14 bits in the key to 10. On larger models this effect is likely to be even more pronounced. In addition the largest observed routing table (without minimisation) went from containing ~1500 entries to ~900 (i.e., too big to fitting).

### CODE REVIEW

**Comment by mundya** [comment]
@tcstewar - if you have time and models that you know otherwise work I'd appreciate it if you could give this branch a try. At the moment the logger prints how successful it's been at reducing the number of different routing keys. I've had SPA sequence running successfully from the GUI.

**Comment by mundya** [comment]
To Do
- [x] Be a bit less aggressive about requiring signals which arrive at the same object to have different keys.

File: nengo_spinnaker/operators/lif.py [hunk 1]
```diff
@@ -1155,7 +1155,7 @@ def get_decoders_and_keys(signals_connections, minimise=False):
         decoders.append(decoder[keep, :])
         for i, k in zip(range(decoder.shape[0]), keep):
             if k:
-                keys.append(signal.keyspace(index=i))
+                keys.append((signal, dict(index=i)))
```
Comment (neworderofjamie) [inline nengo_spinnaker/operators/lif.py]: Maybe be consistent and stick to `{"index": i}`

File: nengo_spinnaker/regions/filters.py [hunk 2]
```diff
@@ -320,17 +322,48 @@ def write_subregion_to_file(self, fp, *args, **kwargs):
...
+                ks.get_value(tag=self.filter_routing_tag),
+                ks.get_mask(tag=self.filter_routing_tag),
+                ks.get_mask(field=self.index_field)
+            )
+
+            if keymask not in keymask_to_targets:
+                keymask_to_targets[keymask] = sorted(
+                    signal_id_to_targets[signal_id])
+            else:
+                assert (
```
Comment (neworderofjamie) [inline nengo_spinnaker/regions/filters.py:354]: Maybe explain what case this is trapping a little more

File: nengo_spinnaker/netlist/key_allocation.py [hunk 3]
```diff
@@ -0,0 +1,224 @@
...
+
+    # Get unique identifiers for each signal
+    signal_ids = assign_mn_net_ids(signal_routes, prior_constraints)
+
+    # Assign keyspaces to the signals
+    for signal, i in iteritems(signal_ids):
+        signal.keyspace = keyspaces["nengo"](connection_id=i)
+
+        # Expand the keyspace to fit the required indices
+        signal.keyspace(index=signal.weight - 1)
```
Comment (neworderofjamie) [inline nengo_spinnaker/netlist/key_allocation.py]: if weight is actually always width, it probably shouldn't be called weight :)

File: nengo_spinnaker/netlist/key_allocation.py [hunk 4]
```diff
@@ -0,0 +1,224 @@
...
+    #            |  /            (f) --- (g)
+    #            | /               \     /
+    #   (a) --- (b)                 \   /
+    #             \                  (h)
+    #              \
+    #               \          (i)
+    #               (e)
+    #
+    # Where a valid colouring would be:
+    #   0: (b), (f), (h)
```
Comment (neworderofjamie) [inline nengo_spinnaker/netlist/key_allocation.py]: i

**Comment by neworderofjamie** [comment]
It is optimistic whether I would spot any algorithmic issues, but looks good!

**Pull request merged**